applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,360 @@
+		}
+		if (op == GPS_START && gps_dev != NULL) {
+			err = at_params_unsigned_short_get(&at_param_list, 2, &interval);
+			if (err < 0) {
+				return err;
+			}
+			(void)at_params_unsigned_int_get(&at_param_list, 3, &timeout);
+			/* GNSS API spec check */
+			if (interval == 0 || interval == 1 ||
+			   (interval >= 10 && interval <= 1800)) {
+				if (interval == 0) {
+					gps_cfg.nav_mode = GPS_NAV_MODE_SINGLE_FIX;
+				} else if (interval == 1) {
+					gps_cfg.nav_mode = GPS_NAV_MODE_CONTINUOUS;
+				} else {
+					gps_cfg.nav_mode = GPS_NAV_MODE_PERIODIC;
+				}
+				gps_cfg.interval   = interval;
+				gps_cfg.timeout    = timeout;
+
+				err = gps_start(gps_dev, &gps_cfg);
+				if (err) {
+					LOG_ERR("Failed to start GPS, error: %d", err);
+				} else {
+					ttft_start = k_uptime_get();
+				}
+			}
+		} else if (op == GPS_STOP) {
+			err = gps_stop(gps_dev);
+		} break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf, "\r\n#XGPS: (%d,%d),<interval>,<timeout>\r\n",
+			GPS_STOP, GPS_START);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XAGPS commands
+ *  AT#XAGPS=<op>[,<interval>[,<timeout>]]
+ *  AT#XAGPS? READ command not supported
+ *  AT#XAGPS=?
+ */
+int handle_at_agps(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+#if defined(CONFIG_SLM_AGPS)
+	uint16_t op;
+	uint16_t interval;
+	uint32_t timeout = 0;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		err = at_params_unsigned_short_get(&at_param_list, 1, &op);
+		if (err < 0) {
+			return err;
+		}
+		if (op == AGPS_START && gps_dev != NULL) {
+			err = at_params_unsigned_short_get(&at_param_list, 2, &interval);
+			if (err < 0) {
+				return err;
+			}
+			(void)at_params_unsigned_int_get(&at_param_list, 3, &timeout);
+			/* GNSS API spec check */
+			if (interval == 0 || interval == 1 ||
+			   (interval >= 10 && interval <= 1800)) {
+				if (interval == 0) {
+					gps_cfg.nav_mode = GPS_NAV_MODE_SINGLE_FIX;
+				} else if (interval == 1) {
+					gps_cfg.nav_mode = GPS_NAV_MODE_CONTINUOUS;
+				} else {
+					gps_cfg.nav_mode = GPS_NAV_MODE_PERIODIC;
+				}
+				gps_cfg.interval   = interval;
+				gps_cfg.timeout    = timeout;
+				/* start with connecting nRF Cloud */
+				err = cloud_connect(nrf_cloud);
+				if (err) {
+					LOG_ERR("Cloud connection failed, error: %d", err);
+				}
+			}
+		} else if (op == AGPS_STOP) {
+			err = cloud_disconnect(nrf_cloud);
+			if (err) {
+				LOG_ERR("Cloud disconnection failed, error: %d", err);
+			}
+			err = gps_stop(gps_dev);
+		} break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf, "\r\n#XAGPS: (%d,%d),<interval>,<timeout>\r\n",
+			AGPS_STOP, AGPS_START);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+#endif /* CONFIG_SLM_AGPS */
+
+	return err;
+}
+
+/**@brief API to initialize GNSS AT commands handler
+ */
+int slm_at_gnss_init(void)
+{
+	int err;
+
+	if (gps_dev == NULL) {
```

Comment:
This NULL check is unnecessary.

Reply:
slm_at_gnss_init() could be called twice, due to SLM support power saving so in case of sleep/wakeup it is called again.

Reply:
It's safe to call `device_get_binding()` again, actually the code for `NRF_CLOUD` initialization below doesn't have the NULL check.

If `slm_at_gnss_init()` can be called twice, I would suggest to check `gps_dev` and `nrf_cloud` and skip the `gps_init()` and `cloud_init()` calls. Both will return an error in case the library has already been initialized, but that would result in an error trace.

Reply:
Updated as advised